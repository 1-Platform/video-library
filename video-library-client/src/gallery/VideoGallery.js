import React, { useEffect, useState } from "react";
import {
  Card,
  CardBody,
  Button,
  Gallery,
  PageSection,
} from "@patternfly/react-core";
import useGlobal from "../GlobalState";
import { PlusCircleIcon } from "@patternfly/react-icons";
import { NavLink } from "react-router-dom";
import VideoAPIs from "../services/VideoAPIs";
import Helpers from "../services/Helpers";

const VideoGallery = (props) => {
  const [pageVideos, setPageVideos] = useState([]);
  const [globalState, globalActions] = useGlobal();
  const { page, perPage, sortBy, searchTerm, videos } = globalState;

  /**
   * Sorter for sorting the list of videos based on the criteria
   */
  const sorter = (a, b) => {
    let result = 0;
    switch (sortBy) {
      case "Name":
        result = a.title.toUpperCase() < b.title.toUpperCase() ? -1 : 1;
        break;
      case "Most viewed":
        result = a.views > b.views ? -1 : 1;
        break;
      case "Newest first":
        result = a.createdOn < b.createdOn ? -1 : 1;
        break;
      case "Oldest first":
        result = a.createdOn < b.createdOn ? 1 : -1;
        break;
      default:
        result = a.title.toUpperCase() > b.title.toUpperCase() ? 1 : -1;
    }
    return result;
  };

  const extractSetPerPageVideos = (videos) => {
    const pageVideos = videos.slice((page - 1) * perPage, page * perPage);
    setPageVideos(pageVideos);
  };

  useEffect(() => {
    const filterProps = ["title", "description", "tags", "createdBy", "source"];
    let sortedVideos = videos;
    if (videos.length === 0) {
      VideoAPIs.listVideos()
        .then((videoList) => {
          globalActions.setVideoCount(videoList.length);
          sortedVideos = videoList.sort(sorter);
          globalActions.setVideos(sortedVideos);
          extractSetPerPageVideos(sortedVideos);
        })
        .catch((err) => {
          if (window.OpNotification) {
            window.OpNotification.danger({
              subject: err.message,
              body: `There was a problem while fetching all the videos. Please try again in sometime.`,
            });
          } else {
            console.error(err);
          }
        });
    } else {
      if (searchTerm) {
        sortedVideos = Helpers.multiPropsFilter(
          videos,
          filterProps,
          searchTerm
        );
        globalActions.setVideoCount(sortedVideos.length);
      }
      sortedVideos.sort(sorter);
      extractSetPerPageVideos(sortedVideos);
      globalActions.setVideoCount(sortedVideos.length);
    }
  }, [page, sortBy, searchTerm, videos]);

  return (
    <React.Fragment>
      <PageSection>
        {pageVideos.length === 0 ? (
          <div>
            <div className="no-video-msg">
              No videos found for the selected filters.
            </div>
            <Button
              variant="primary"
              icon={<PlusCircleIcon />}
              onClick={props.openAddModal}
            >
              {" "}
              Add Video
            </Button>
          </div>
        ) : (
          <Gallery hasGutter className="video-gallery">
            {pageVideos.map((video, key) => (
              <React.Fragment key={key}>
                <Card className="card" isHoverable isSelectable isFlat>
                  <CardBody>
                    <section className="video-info">
                      <h5>
                        <NavLink
                          to={{
                            pathname: `/watch/${video._id}`,
                            key: video._id,
                            state: { video: video },
                          }}
                          title="Watch Video"
                        >
                          {video.title || "Title not available"}
                        </NavLink>
                      </h5>
                      <p className="uploaded-by">
                        {video.createdBy || "No owner"}
                      </p>
                      <NavLink
                        className="edit-video-link"
                        to={{
                          pathname: `/edit/${video._id}`,
                          key: video._id,
                          state: { video: video },
                        }}
                        title="Edit Video"
                      >
                        <ion-icon name="create-outline"></ion-icon>
                      </NavLink>
                      <div className="footer">
                        <span>{Helpers.formatDate(video.createdOn)}</span>
                        <span className="views">
                          {Helpers.roundOffViews(video.views || 0)} views
                        </span>
                      </div>
                    </section>
                    <aside>
                      <NavLink
                        to={{
                          pathname: `/watch/${video._id}`,
                          key: video._id,
                          state: { video: video },
                        }}
                      >
                        <ion-icon
                          name="play-circle-outline"
                          title="Watch Video"
                        ></ion-icon>
                      </NavLink>
                      <div className="info">
                        <ion-icon
                          name={Helpers.getSourceIcon(video.source)}
                          title={`Video Source: ${video.source.toUpperCase()}`}
                        ></ion-icon>
                        <span className="length">
                          {video.length
                            ? Helpers.humanizeTime(video.length)
                            : video.approxLength
                            ? video.approxLength.toLowerCase()
                            : "N/A"}
                        </span>
                      </div>
                    </aside>
                  </CardBody>
                </Card>
              </React.Fragment>
            ))}
          </Gallery>
        )}
      </PageSection>
    </React.Fragment>
  );
};

export default VideoGallery;
