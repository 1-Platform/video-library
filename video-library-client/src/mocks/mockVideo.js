export const videos = {
  data: {
    listVideos: [
      {
        approxLength: "< 5 min",
        categories: [],
        createdBy: {
          name: "Spider Man",
          rhatUUID: "0970b4a0-ee83-22e7-72a6-005a7a0b44d",
        },
        createdOn: "2019-07-17T21:06:59.000Z",
        description: "This is a test video.",
        //fileID: "5d2f8df499410e01d3da3d62",
        length: null,
        mailingLists: [],
        //source: "mojo",
        tags: ["tag1", "tag2"],
        title: "Test video",
        updatedBy: {
          name: "Ghanshyam Lohar",
          rhatUUID: "41c96546-92ec-11ea-af27-0a67ac234eca",
        },
        updatedOn: "2019-08-22T11:43:50.000Z",
        videoURL: "",
        views: 3,
        _id: "5d2f8df499410e01d3da4f34",
      },
    ],
  },
};
