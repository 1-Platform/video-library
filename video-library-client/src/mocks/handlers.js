// src/mocks/handlers.js
import { rest } from "msw";
import { videos } from "./mockVideo";
export const handlers = [
  // Handles a PST /login request
  rest.post("http://localhost:8080/graphql", async (req, res, ctx) => {
    return res(ctx.status(200));
  }),
];
