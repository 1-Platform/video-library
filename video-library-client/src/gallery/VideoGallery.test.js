import React from "react";
import {
  render as rtlRender,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import VideoGallery from "./VideoGallery";
import { videos } from "../mocks/mockVideo";
import { rest } from "msw";
import { server } from "../mocks/server";
import renderer from "react-test-renderer";

test("renders video list on home", async () => {
  server.use(
    rest.post("http://localhost:8080/graphql", async (req, res, ctx) => {
      return res(ctx.status(200), ctx.json(videos));
    })
  );
  const openAddModal = jest.fn();
  rtlRender(
    <MemoryRouter>
      <VideoGallery openAddModal={openAddModal} />
    </MemoryRouter>
  );

  await waitForElementToBeRemoved(() => screen.queryByText("Loading"));
  const videoTitle = screen.getByText("Test video");
  expect(videoTitle).toBeInTheDocument();
});

test("VideoGallery renders correctly", () => {
  server.use(
    rest.post("http://localhost:8080/graphql", async (req, res, ctx) => {
      return res(ctx.status(200), ctx.json(videos));
    })
  );
  const openAddModal = jest.fn();
  const tree = renderer
    .create(
      <MemoryRouter>
        <VideoGallery openAddModal={openAddModal} />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
