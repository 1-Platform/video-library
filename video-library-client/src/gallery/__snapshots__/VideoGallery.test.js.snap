// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoGallery renders correctly 1`] = `
<section
  className="pf-c-page__main-section video-page-section"
>
  <div
    className="pf-c-empty-state"
  >
    <div
      className="pf-c-empty-state__content"
    >
      <div
        className="pf-c-empty-state__icon"
      >
        <span
          aria-valuetext="Loading..."
          className="pf-c-spinner pf-m-xl"
          role="progressbar"
        >
          <span
            className="pf-c-spinner__clipper"
          />
          <span
            className="pf-c-spinner__lead-ball"
          />
          <span
            className="pf-c-spinner__tail-ball"
          />
        </span>
      </div>
      <h4
        className="pf-c-title pf-m-lg"
      >
        Loading
      </h4>
    </div>
  </div>
  <div
    className="pf-l-gallery pf-m-gutter video-gallery"
  />
</section>
`;
